app/services/process_payment_intent.rb

```diff
@@ -50,25 +49,36 @@ def call!
 
   private
 
-  attr_reader :order, :payment_intent, :last_payment
+  attr_reader :order, :payment_intent, :payment
 
-  def valid?
-    order.present? && matches_last_payment?
+  def process_payment
+    if order.state == "payment"
+      # Moves the order to completed, which calls #process_payments!
+      OrderWorkflow.new(order).next
+    else
+      order.process_payments!
+    end
+  end
```

Comment:
This is IMO flagging a flaw. So far it was OrderWorkflow the one responsible for this, and this change leaks this concern. As a result, there are now two places we will need to check next time we're investigating or developing something related to payment processing. I would suggest we create a tech debt issue and carefully investigate this design problem without any rush.

I think the underlying issue here is having all payment processing logic hanging from the order model. I always knew this wasn't a good idea and I think now I have proof.

Reply:
Agreed it needs a lot of re-thinking! The issue here is `#process_payments!` only gets called on the final state transition to complete, and there are lots of cases (including subscriptions) where the order is already compete when we want to process the payment via Stripe.

Reply:
I see your comment on line 56 and it makes sense to me now. It says `OrderWorkflow` does two things: moving to completion && calling process payments => this violates SRP. If these two weren't glued together we could choose and process payments always, regardless of the order state.

Reply:
A lot of the issues here come from the fact that the mechanics of the checkout order placement flow and the admin order placement flow are totally different, but we try to re-use the same logic as if there are no differences. There's lots of things that need to be decoupled and updated to make it work nicely with both!